### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by greyhillman**
This adds some documentation to the project.

This is a work-in-progress. This is mostly to be used as feedback on what to add and changes.

### CODE REVIEW

**Comment by greyhillman** [comment]
Add a section how how to contribute to the developer docs.

**Comment by greyhillman** [comment]
Add a tutorial on how to run the model in a literate programming way.

File: docs/glossary.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#### Capacity
+
+A variable essentially. You can reference this capacity in certain places in 
+order to introduce a variable amount of something.
+
+For example, you can create a capacity in the capacities tab and then reference
+it in the capacity row in the Converters tab. This makes it so that it finds 
+the most optimal converter capacity as well as the other variables.
+
+## Civil Engineering
```
Comment (greyhillman) [inline docs/glossary.md]: Change this to Energy Hub Model.

File: docs/index.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Documentation
+
+## User Installation
+
+A tutorial on how to install on Linux-based systems can be found
+[here](install/linux.sh)
```
Comment (greyhillman) [inline docs/index.md]: Add period.

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+## Command-line Usage
+
+Can be done via
```
Comment (greyhillman) [inline docs/index.md]: Change to `Run:`

File: docs/index.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+```bash
+python3.6 cli.py --help
+```
```
Comment (greyhillman) [inline docs/index.md]:
```
to see how to use the program via the command-line.
Also, look at the top of the cli.py program for more up-to-date details.
```

File: docs/index.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+```bash
+python3.6 cli.py
+```
+
+This will print out the results to the command line and also put the results in
+the output file specified in the `config.yaml` file.
+
+## Glossary
+
+A glossary of terms can be found [here](glossary.md)
```
Comment (greyhillman) [inline docs/index.md]: Add period.

File: docs/index.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+## Developer Guide
+
+Can be found [here](dev_guide.md)
```
Comment (greyhillman) [inline docs/index.md]: Add period.

File: docs/dev_guide.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# Developer Guide
+
+This document contains information for extending and modifying the Energy Hub 
+Model.
+
+For more concrete API details, see the code. This will be more up-to-date than
```
Comment (greyhillman) [inline docs/dev_guide.md]: That will be

File: docs/dev_guide.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# Developer Guide
+
+This document contains information for extending and modifying the Energy Hub 
+Model.
+
+For more concrete API details, see the code. This will be more up-to-date than 
+this or any other documentation.
```
Comment (greyhillman) [inline docs/dev_guide.md]: Remove `this or`

File: docs/dev_guide.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+We'll review your changes, make suggestions, and merge it when it's good.
+
+### Non-core Development
+
+If you want to extend the functionality of the model, and you don't have push
+access to the repo, create a fork of the project on GitHub.
+Then develop on your fork as you wish.
+If you think your changes might be useful to the main core, create a pull
+request from your fork to the repo.
+Then follow the same instructions as for core development: add a main
```
Comment (greyhillman) [inline docs/dev_guide.md]: `Then follow the same instructions as for core development.`
Remove the rest

File: docs/dev_guide.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+collaborator as a reviewer.
+From there, it's the same as the core development.
+
+## Tutorials
```
Comment (greyhillman) [inline docs/dev_guide.md]: `Tutorials/Examples`

File: docs/glossary.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+- [Pyomo](#pyomo)
+- [Request Format/Excel Sheet](#request-formatexcel-sheet)
+- [Civil Engineering](#civil-engineering)
+
+## Pyomo
+
+#### Variable
+
+Much like a mathematical variable.
```
Comment (greyhillman) [inline docs/glossary.md]: `Some mathematical variable you want to solve for.`

File: docs/glossary.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+This is usually something we want to solve for.
+
+#### Parameter
+
+Much like a mathematical parameter.
```
Comment (greyhillman) [inline docs/glossary.md]: Some mathematical constant that is passed into the model.

File: docs/glossary.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+This is some constant that should not change.
+
+#### Constraint
+
+Much like a mathematical constraint.
```
Comment (greyhillman) [inline docs/glossary.md]: An equation that relates variables and parameters; the same as a mathematical constraint.

File: docs/glossary.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+An equation that relates variables and parameters.
```
Comment (greyhillman) [inline docs/glossary.md]: Add a definition for `ConstantOrVar`.
```
Some attribute of something that can either be a constant value (like a parameter) or a Pyomo variable.
```

Comment (greyhillman) [inline docs/glossary.md]: Actually no. This is not related to Pyomo.

File: docs/glossary.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+or Electricity is considered a stream.
+
+#### Converter
+
+A piece of technology or machine that converts one or more input energy streams 
+into one or more output energy streams.
+
+#### Storage
+
+A piece of technology that stores an energy stream.
```
Comment (greyhillman) [inline docs/glossary.md]: Example: A battery or hot water tank.

File: docs/index.md (hunk 3 above)
Comment (greyhillman) [inline docs/index.md]: Actually remove that and just have the command.

File: docs/index.md (hunk 4 above)
Comment (greyhillman) [inline docs/index.md]: Remove the first line of above comment.

File: docs/index.md [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+Can be done via
+
+```bash
+python3.6 cli.py --help
+```
+
+## How to run a Model
+
+Given that you have a proper Excel file, first edit the `config.yaml` file to
+point to your file. See the `config.yaml` for more detail.
```
Comment (greyhillman) [inline docs/index.md]: `details`

File: docs/install/linux.sh [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+# Now we switch back into `python-ehub`
+cd ..
+cd python-ehub
+
+# Now that we have the programs to run the Energy Hub Model, we install all the
+# libraries it uses:
+python3.6 -m pip install -r requirements.txt
+
+# That should install all the necessary libraries to run the Energy Hub Model.
```
Comment (greyhillman) [inline docs/install/linux.sh]: Remove. Says this above

File: docs/tutorials/extension.py [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    Here, we can add our own constraints.
+    """
+
+    @staticmethod
+    @constraint()
+    def new_constraint(model):
+        return model.TankSize + model.BatSize <= 200
+
+current_directory = os.path.dirname(os.path.realpath(__file__))
+model = MyModel(excel=f"{current_directory}/extension.xlsx")
```
Comment (greyhillman) [inline docs/tutorials/extension.py]: Make this line cross-platform

File: docs/tutorials/extension.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+# noinspection PyShadowingNames
+class MyModel(EHubModel):
+    """
+    This is a subclass of EHubModel.
+
+    Here, we can add our own constraints.
+    """
+
+    @staticmethod
+    @constraint()
```
Comment (greyhillman) [inline docs/tutorials/extension.py]:
```python
@constraint()  # Marks this function as a constraint of the model
```

Comment (greyhillman) [inline docs/tutorials/extension.py]: I've expanded more in the docstring instead.

**Comment by greyhillman** [comment]
Add documentation on how server works. Include some info on Docker.

**Comment by greyhillman** [comment]
That might be more appropriate for the PR of the docker server, but here it's fine.

**Comment by greyhillman** [comment]
@wesleyb321 @ralphevins 
You guys mind taking a look at this sometime?
I'd like some suggestions on what things you would like to know and improvements on what it already here.

**Comment by greyhillman** [comment]
Add example of updating request format schema/excel file.

**Comment by greyhillman** [comment]
Add something about style guidelines (docstring format is Google format).

File: docs/server.md [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+To gracefully stop:
+```bash
+docker-compose down
+```
+
+To kill it with fire:
+```bash
+docker-compose kill
+```
```
Comment (greyhillman) [inline docs/server.md]: Add section on how to build and transfer the image to another computer.

File: docs/tutorials/extension.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
+"""
+This is a file for a tutorial on how to extend the model without modifying it.
+
+To run this file, from the root directory of this project, do
+
+    python3.6 -m docs.tutorials.extension
+"""
+import os
+
+from cli import pretty_print
```
Comment (greyhillman) [inline docs/tutorials/extension.py]: Rebase this file.

**Comment by greyhillman** [comment]
Also add section on common Docker commands.

File: docs/server.md (hunk 20 above)
Comment (greyhillman) [inline docs/server.md]: Done.

File: docs/docker.md [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+You want to start a container from the image `test` with bash
+
+```bash
+docker run -it test /bin/bash
+```
+
+# Docker-compose
+
+This is a program that simplifies working with `docker` commands.
+`docker` commands can get quite verbose (just run `man docker run` to see what
```
Comment (greyhillman) [inline docs/docker.md]: just run `man docker run`

File: docs/docker.md [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+I mean).
+
+The `docker-compose.yaml` file is a configuration file for running and building
+a docker container or iamge respectively.
```
Comment (greyhillman) [inline docs/docker.md]: image

File: docs/docker.md [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+docker-compose build
+```
+
+## Run the server image
+
+```bash
+docker-compose up -d
+```
+
+The `-d` part means run the container in the background.'
```
Comment (greyhillman) [inline docs/docker.md]: Remove trailing '

File: docs/docker.md [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+
+## Stop the server image
```
Comment (greyhillman) [inline docs/docker.md]: server container

File: docs/explanation.md [hunk 26]
```diff
@@ -1,5 +1,8 @@
 # Explanation of variables and parameters
 
+**Note: This document will get out of date quickly. See the code for more 
+up-to-date explanation.**
```
Comment (greyhillman) [inline docs/explanation.md]: I might just delete this file. It's not really good and it will go out of date quickly.
To top it off, it's for the Jupyter notebook files and not the real Python code.

File: docs/glossary.md [hunk 27]
```diff
@@ -0,0 +1,97 @@
+# Glossary
+
+This file contains terminology that is used throughout the project.
+
+## Table of Contents
+
+- [Pyomo](#pyomo)
```
Comment (greyhillman) [inline docs/glossary.md]: Change this to linear programming.
That way it's more generic and won't go out of date.

File: docs/glossary.md [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+- [Request Format/Excel Sheet](#request-formatexcel-sheet)
+- [Civil Engineering](#civil-engineering)
+
+## Pyomo
```
Comment (greyhillman) [inline docs/glossary.md]: Include examples in this section.

File: docs/glossary.md [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+
+Short for photovoltaic.
+Usually used to refer to solar panels.
+
+#### Solar
+
+Anything that uses the sun.
+Does **not** have to refer to solar panels.
+
+#### Capacity
```
Comment (greyhillman) [inline docs/glossary.md]: Need help with this.

File: docs/index.md [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+
+## User Installation
+
+A tutorial on how to install on Linux-based systems can be found
+[here](install/linux.sh).
+
+## Command-line Usage
+
+```bash
+python3.6 cli.py --help
```
Comment (greyhillman) [inline docs/index.md]: Change `cli.py` to `run.py` for the docs.

File: docs/style_guide.md [hunk 31]
```diff
@@ -0,0 +1,24 @@
+# Style Guide
+
+Follow PEP 8 guidelines unless otherwise noted below.
+
+`pylint` also is a good way to ensure that code is up to the style guidelines.
```
Comment (greyhillman) [inline docs/style_guide.md]: Change this to
```
Run `python3.6 ci.py` in order to run all the tests and check for any style problems.
```

File: docs/explanation.md (hunk 26 above)
Comment (ralphevins) [inline docs/explanation.md]: I'd like to keep a version of it that's up to date with the latest code, as it would be a useful reference and I don't see the core code changing very often. Might merge with the glossary.

Comment (greyhillman) [inline docs/explanation.md]: The document itself is more like giving better names to the variables. Anyone should be doing that already; if you have a better name for a variable (ie make it better name what the values reporesent) then change it. That's why I changed stuff like `Yon` to `is_on` and `P` to `energy_imported`.

That's why I want to remove it.

Comment (greyhillman) [inline docs/explanation.md]: To have a reference for what they represent at a higher-level, like how they relate to each other, then I agree. I'd still like to change the names of the variables to something more accurate of what they represent though.

I just don't like having documents like this one that depend on the code heavily. Since if you change the code and not the document, anyone reading the document will get frustrated or confused. In this specific example, it relies on Boran's Jupyter notebook file and not the regular Python code that is in this branch. Also, it relies on Pyomo which has been replaced.

File: docs/glossary.md [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+inequality constraints.
+
+#### Objective Function
+
+This is the function we want to optimize.
+By convention it is always to be minimized.
+
+#### Variable
+
+The set of variables you want to solve for.
```
Comment (greyhillman) [inline docs/glossary.md]: I would have it refer to a single variable. This also uses the definition inside of itself.

I'd have something like
```
A mathematical variable you want to solve for.

Example:
Like the `x` in `x + 3 = 5`.
```

File: docs/glossary.md [hunk 33]
```diff
@@ -0,0 +1,151 @@
...
+- `p >= 1`
+- `c + p >= 5`
+- `c > p`
+
+Now you hopefully have an idea of what an objective function, a variable, and a
+constraint is in the context of linear programming.
+
+## Energy Hub Model
+
+The energy hub model is a form of linear programme that describes an energy
```
Comment (greyhillman) [inline docs/glossary.md]: I'm confused reading this.
I think the grammar is weird. The section between `the energy hub model` and `describes a conversion` is the problem.

Maybe say it is a linear programming problem for converting and storing energy over time.

**Comment by greyhillman** [comment]
Merged so that new documentation can be added. Also to make adding things to the docs can be done in other pull requests.

**Comment by greyhillman** [comment]
Should have squashed the commits. My bad.
